flight/goto.py

```diff
@@ -0,0 +1,92 @@
+    Parameters
+    ----------
+    drone: System
+        a drone object that has all offboard data needed for computation
+    latitude: float
+        a float containing the requested latitude to move to
+    longitude: float
+        a float containing the requested longitude to move to
+    altitude: float
+        a float contatining the requested altitude to go to (in feet)
+    fast_mode: bool
+        a boolean that determines if the drone will take less time checking its precise location
+        before moving on to another waypoint. If its true, it will move faster, if it is false,
+        it will move at normal speed
+    """
+
+    # converts feet into meters
+    altitude = altitude * 0.3048
+
+    # get current altitude
+    async for terrain_info in drone.telemetry.home():
+        absolute_altitude: float = terrain_info.absolute_altitude_m
+        break
+
+    await drone.action.goto_location(latitude, longitude, altitude + absolute_altitude, 0)
+    location_reached: bool = False
+    # First determine if we need to move fast through waypoints or need to slow down at each one
+    # Then loops until the waypoint is reached
+    if fast_mode == True:
+        while not location_reached:
+            logging.info("Going to waypoint")
+            async for position in drone.telemetry.position():
+                # continuously checks current latitude, longitude and altitude of the drone
+                drone_lat: float = position.latitude_deg
+                drone_long: float = position.longitude_deg
+                drone_alt: float = position.relative_altitude_m
+
+                #  accurately checks if location is reached and stops for 15 secs and then moves on.
+                if (
+                    (round(drone_lat, 5) == round(latitude, 5))
+                    and (round(drone_long, 5) == round(longitude, 5))
+                    and (round(drone_alt, 1) == round(altitude, 1))
+                ):
+                    location_reached = True
+                    logging.info("arrived")
+                    asyncio.sleep(15)
+                    break
+
+            # tell machine to sleep to prevent contstant polling, preventing battery drain
+            await asyncio.sleep(1)
+    else:
+        while not location_reached:
+            logging.info("Going to waypoint")
+            async for position in drone.telemetry.position():
+                # continuously checks current latitude, longitude and altitude of the drone
+                drone_lat = position.latitude_deg
+                drone_long = position.longitude_deg
+                drone_alt = position.relative_altitude_m
+
+                # accurately checks if location is reached and stops for 15 secs and then moves on.
+
+                if (
+                    (round(drone_lat, 6) == round(latitude, 6))
+                    and (round(drone_long, 6) == round(longitude, 6))
+                    and (round(drone_alt, 1) == round(altitude, 1))
```

Comment:
The fast mode and normal mode are basically the same here aside from these statements. You could probably condense this to just change the numbers that you need rather than having a bunch of duplicate code under the if/else

Reply:
Just spitballin' here, you could change the fast mode argument to a float, where it would act as a multiplier to the rounding precision, like round(drone_lat, 6*fast_param). For slow mode make it 0, and for fast mode make it greater than 0 but less than 1.